[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62346

Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30003835](https://our.internmc.facebook.com/intern/diff/D30003835)

supriyar committed:
[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize
Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29968871).

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29968871](https://our.internmc.facebook.com/intern/diff/D29968871)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29968871](https://our.internmc.facebook.com/intern/diff/D29968871)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29968871](https://our.internmc.facebook.com/intern/diff/D29968871)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -273,8 +273,9 @@ def __init__(
         assert isinstance(
-            self.activation_post_process, MovingAverageMinMaxObserver
-        ), "Fused observer+fake_quant module only works with MovingAverageMinMaxObserver"
+            self.activation_post_process, MovingAverageMinMaxObserver) or isinstance(
```
optional syntax sugar: `isinstance(self.activation_post_process, (MovingAverageMinMaxObserver, MovingAveragePerChannelMinMaxObserver))`

vkuzo commented on aten/src/ATen/native/quantized/cpu/fused_obs_fake_quant.cpp:
```diff
@@ -146,6 +145,14 @@ std::tuple<at::Tensor, at::Tensor> fused_moving_avg_obs_fake_quant_cpu(
     bool symmetric_quant) {
   // Calculate min/max
   auto observe = observer_on.item().toInt();
+  int64_t size = per_row_fake_quant ? self.size(0) : 1;
```
nit: might be good to add a comment what this code block is doing, for future readers

vkuzo commented on aten/src/ATen/native/quantized/cpu/fused_obs_fake_quant.cpp:
```diff
@@ -146,6 +145,14 @@ std::tuple<at::Tensor, at::Tensor> fused_moving_avg_obs_fake_quant_cpu(
+  if (per_row_fake_quant && running_min.numel() == 0) {
+    float inf = std::numeric_limits<float>::infinity();
+    running_min.resize_(size).fill_(inf);
+    running_max.resize_(size).fill_(inf);
```
n00b question, should this be `-inf` instead of `inf`?

supriyar replied:
Yes, ideally. Although for the op it doesn't make a difference because we check for `std::isinf` which returns true either way.

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29968871](https://our.internmc.facebook.com/intern/diff/D29968871)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29968871](https://our.internmc.facebook.com/intern/diff/D29968871)

[ghstack-poisoned]

HDCharles reviewed (approved):
looks good thoguh it looks like if fused_obs_fake_quant.cpp has an explanation comment, fused_obs_fake_quant.cpp should have one as well.

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29968871](https://our.internmc.facebook.com/intern/diff/D29968871)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29968871).

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29968871](https://our.internmc.facebook.com/intern/diff/D29968871)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30003835).

supriyar committed:
Update on "[quant] Support PerChannel quantization in FusedMovingAvgObsFakeQuantize"


Summary:
Update the operator code to resize the min/max tensors if per-channel quant is selected. We need to do this because by default the observer creates empty tensors for min/max and scale/zero_point values when per-channel quantization is enabled

Test Plan:
python test/test_quantization.py test_fused_mod_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30003835](https://our.internmc.facebook.com/intern/diff/D30003835)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30003835).
